File: src/c/ch01/1.3/l2/extgcd_inv.c
#include "extgcd_inv.h"
#include <stdlib.h>

int64_t extended_gcd(int64_t a, int64_t b, int64_t *x, int64_t *y) {
    int64_t old_r = a, r = b;
    int64_t old_s = 1, s = 0;
    int64_t old_t = 0, t = 1;
    while (r != 0) {
        int64_t q = old_r / r;
        int64_t tmp;

        tmp = old_r - q * r; old_r = r; r = tmp;
        tmp = old_s - q * s; old_s = s; s = tmp;
        tmp = old_t - q * t; old_t = t; t = tmp;
    }
    if (x) *x = old_s;
    if (y) *y = old_t;
    return old_r; /* gcd */
}

bool invmod_i64(int64_t a, int64_t m, int64_t *inv) {
    if (m <= 0) return false;
    int64_t x, y;
    int64_t g = extended_gcd(a, m, &x, &y);
    if (g != 1 && g != -1) return false;
    /* Normalize into [0, m-1] */
    int64_t res = x % m;
    if (res < 0) res += m;
    if (inv) *inv = res;
    return true;
}

File: archived/src/erlang/ch01/1.3/l2/src/extgcd_inv.erl
-module(extgcd_inv).
-export([extended_gcd/2, invmod/2]).

%% extended_gcd(A,B) -> {G,X,Y} with A*X + B*Y = G = gcd(A,B).
extended_gcd(A, B) ->
    extended_gcd_(A, B, 1, 0, 0, 1).

extended_gcd_(OldR, R, OldS, S, OldT, T) when R =:= 0 ->
    {OldR, OldS, OldT};
extended_gcd_(OldR, R, OldS, S, OldT, T) ->
    Q = OldR div R,
    extended_gcd_(R,
                  OldR - Q*R,
                  S,
                  OldS - Q*S,
                  T,
                  OldT - Q*T).

%% invmod(A, M) -> X in 0..M-1 with (A*X) rem M = 1; error if no inverse.
invmod(A, M) when M =< 0 ->
    erlang:error(badarg);
invmod(A, M) ->
    {G, X, _Y} = extended_gcd(A, M),
    case (G =:= 1) orelse (G =:= -1) of
        true  -> ((X rem M) + M) rem M;
        false -> erlang:error({no_inverse, {A, M}})
    end.